templates/Containerfile

```diff
@@ -11,6 +12,12 @@ FROM scratch as stage-modules
 COPY --from=ghcr.io/ublue-os/bling:latest /modules /modules
 COPY ./modules /modules
 
+# This stage is responsible for collecting all akmods rpms.
+FROM scratch as stage-akmods
+{%- for image in self::get_akmods_image_list(recipe.modules_ext, os_version) %}
+COPY --from=ghcr.io/ublue-os/{{ image }} /rpms /rpms
```

Comment:
Wouldn't this overwrite the RPMs each time a different set of akmods is specified, always using the last one for all the steös, as all akmods images have all akmods?

Reply:
I notice that in Containerfile, Bazzite installs regular akmods first & then it installs nvidia akmod.

Here's how they do it (took from main nvidia's install.sh & post-install.sh script, which Bazzite uses):

```
rpm-ostree install \
    /tmp/akmods-rpms/ublue-os/ublue-os-nvidia-addons-*.rpm

source /tmp/akmods-rpms/kmods/nvidia-vars.${NVIDIA_MAJOR_VERSION}

if [[ "${IMAGE_NAME}" == "kinoite" ]]; then
    VARIANT_PKGS="supergfxctl-plasmoid"
elif [[ "${IMAGE_NAME}" == "silverblue" ]]; then
    VARIANT_PKGS="gnome-shell-extension-supergfxctl-gex"
else
    VARIANT_PKGS=""
fi

rpm-ostree install \
    xorg-x11-drv-${NVIDIA_PACKAGE_NAME}-{,cuda-,devel-,kmodsrc-,power-}${NVIDIA_FULL_VERSION} \
    xorg-x11-drv-${NVIDIA_PACKAGE_NAME}-libs.i686 \
    nvidia-container-toolkit nvidia-vaapi-driver supergfxctl ${VARIANT_PKGS} \
    /tmp/akmods-rpms/kmods/kmod-${NVIDIA_PACKAGE_NAME}-${KERNEL_VERSION}-${NVIDIA_AKMOD_VERSION}.fc${RELEASE}.rpm
    
sed -i 's@enabled=1@enabled=0@g' /etc/yum.repos.d/{eyecantcu-supergfxctl,nvidia-container-toolkit}.repo

systemctl enable ublue-nvctk-cdi.service
semodule --verbose --install /usr/share/selinux/packages/nvidia-container.pp

if [[ "${IMAGE_NAME}" == "sericea" ]]; then
    mv /etc/sway/environment{,.orig}
    install -Dm644 /usr/share/ublue-os/etc/sway/environment /etc/sway/environment
fi    
```

Reply:
> all the steös, as all akmods images have all akmods

Ok I can go even further and give each set their own stage and use that stage for the corresponding RUN's

Reply:
Ok so latest update creates this Containerfile:

```dockerfile
FROM scratch as stage-config
COPY ./config /config

# Copy modules
# The default modules are inside ublue-os/bling
# Custom modules overwrite defaults
FROM scratch as stage-modules
COPY --from=ghcr.io/ublue-os/bling:latest /modules /modules
COPY ./modules /modules
FROM scratch as stage-akmods-surface-545
COPY --from=ghcr.io/ublue-os/akmods:surface-39 /rpms /rpms
COPY --from=ghcr.io/ublue-os/akmods-nvidia:surface-39-545 /rpms /rpms
FROM scratch as stage-akmods-main
COPY --from=ghcr.io/ublue-os/akmods:main-39 /rpms /rpms

# This stage is responsible for holding onto
# exports like the exports.sh
FROM docker.io/alpine as stage-exports
RUN printf "#!/usr/bin/env bash\n\nget_yaml_array() { \n  readarray -t \"\$1\" < <(echo \"\$3\" | yq -I=0 \"\$2\")\n} \n\nexport -f get_yaml_array\nexport OS_VERSION=\$(grep -Po '(?<=VERSION_ID=)\d+' /usr/lib/os-release)" >> /exports.sh && chmod +x /exports.sh

FROM ghcr.io/ublue-os/silverblue-surface:39

LABEL org.blue-build.build-id="895eb408-c23a-4172-b17d-0049d2fc85d6"
LABEL org.opencontainers.image.title="template"
LABEL org.opencontainers.image.description="This is my personal OS image."
LABEL io.artifacthub.package.readme-url=https://raw.githubusercontent.com/blue-build/cli/main/README.md

ARG RECIPE=./config/recipe.yml
ARG IMAGE_REGISTRY=localhost

ARG CONFIG_DIRECTORY="/tmp/config"
ARG IMAGE_NAME="template"
ARG BASE_IMAGE="ghcr.io/ublue-os/silverblue-surface"
RUN \
  --mount=type=tmpfs,target=/tmp \
  --mount=type=tmpfs,target=/var \
  --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
  --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
  --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
  --mount=type=bind,from=stage-akmods-surface-545,src=/rpms,dst=/tmp/rpms,rw \
  --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
  --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
  chmod +x /tmp/modules/akmods/akmods.sh \
  && source /tmp/exports.sh && /tmp/modules/akmods/akmods.sh '{"type":"akmods","base":"surface","nvidia-version":545,"install":["nvidia","openrazer"]}'
RUN \
  --mount=type=tmpfs,target=/tmp \
  --mount=type=tmpfs,target=/var \
  --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
  --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
  --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
  --mount=type=bind,from=stage-akmods-main,src=/rpms,dst=/tmp/rpms,rw \
  --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
  --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
  chmod +x /tmp/modules/akmods/akmods.sh \
  && source /tmp/exports.sh && /tmp/modules/akmods/akmods.sh '{"type":"akmods","install":["openrgb"]}'
COPY ./config/files/usr /usr
RUN \
  --mount=type=tmpfs,target=/tmp \
  --mount=type=tmpfs,target=/var \
  --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
  --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
  --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
  --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
  --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
  chmod +x /tmp/modules/script/script.sh \
  && source /tmp/exports.sh && /tmp/modules/script/script.sh '{"type":"script","scripts":["example.sh"]}'
RUN \
  --mount=type=tmpfs,target=/tmp \
  --mount=type=tmpfs,target=/var \
  --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
  --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
  --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
  --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
  --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
  chmod +x /tmp/modules/rpm-ostree/rpm-ostree.sh \
  && source /tmp/exports.sh && /tmp/modules/rpm-ostree/rpm-ostree.sh '{"type":"rpm-ostree","repos":["https://copr.fedorainfracloud.org/coprs/atim/starship/repo/fedora-%OS_VERSION%/atim-starship-fedora-%OS_VERSION%.repo"],"install":["micro","starship"],"remove":["firefox","firefox-langpacks"]}'
RUN \
  --mount=type=tmpfs,target=/tmp \
  --mount=type=tmpfs,target=/var \
  --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
  --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
  --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
  --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
  --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
  chmod +x /tmp/modules/default-flatpaks/default-flatpaks.sh \
  && source /tmp/exports.sh && /tmp/modules/default-flatpaks/default-flatpaks.sh '{"type":"default-flatpaks","notify":true,"system":{"install":["org.mozilla.firefox","org.gnome.Loupe","one.ablaze.floorp//lightning"],"remove":["org.gnome.eog"]}}'
RUN \
  --mount=type=tmpfs,target=/tmp \
  --mount=type=tmpfs,target=/var \
  --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
  --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
  --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
  --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
  --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
  chmod +x /tmp/modules/signing/signing.sh \
  && source /tmp/exports.sh && /tmp/modules/signing/signing.sh '{"type":"signing"}'
RUN \
  --mount=type=tmpfs,target=/tmp \
  --mount=type=tmpfs,target=/var \
  --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
  --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
  --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
  --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
  --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
  chmod +x /tmp/modules/test-module/test-module.sh \
  && source /tmp/exports.sh && /tmp/modules/test-module/test-module.sh '{"type":"test-module"}'

COPY --from=gcr.io/projectsigstore/cosign /ko-app/cosign /usr/bin/cosign
COPY --from=ghcr.io/blue-build/cli:latest-installer /out/bluebuild /usr/bin/bluebuild

RUN ostree container commit
```

Reply:
> Ok so latest update creates this Containerfile:
> 
> ```dockerfile
> FROM scratch as stage-config
> COPY ./config /config
> 
> # Copy modules
> # The default modules are inside ublue-os/bling
> # Custom modules overwrite defaults
> FROM scratch as stage-modules
> COPY --from=ghcr.io/ublue-os/bling:latest /modules /modules
> COPY ./modules /modules
> FROM scratch as stage-akmods-surface-545
> COPY --from=ghcr.io/ublue-os/akmods:surface-39 /rpms /rpms
> COPY --from=ghcr.io/ublue-os/akmods-nvidia:surface-39-545 /rpms /rpms
> FROM scratch as stage-akmods-main
> COPY --from=ghcr.io/ublue-os/akmods:main-39 /rpms /rpms
> 
> # This stage is responsible for holding onto
> # exports like the exports.sh
> FROM docker.io/alpine as stage-exports
> RUN printf "#!/usr/bin/env bash\n\nget_yaml_array() { \n  readarray -t \"\$1\" < <(echo \"\$3\" | yq -I=0 \"\$2\")\n} \n\nexport -f get_yaml_array\nexport OS_VERSION=\$(grep -Po '(?<=VERSION_ID=)\d+' /usr/lib/os-release)" >> /exports.sh && chmod +x /exports.sh
> 
> FROM ghcr.io/ublue-os/silverblue-surface:39
> 
> LABEL org.blue-build.build-id="895eb408-c23a-4172-b17d-0049d2fc85d6"
> LABEL org.opencontainers.image.title="template"
> LABEL org.opencontainers.image.description="This is my personal OS image."
> LABEL io.artifacthub.package.readme-url=https://raw.githubusercontent.com/blue-build/cli/main/README.md
> 
> ARG RECIPE=./config/recipe.yml
> ARG IMAGE_REGISTRY=localhost
> 
> ARG CONFIG_DIRECTORY="/tmp/config"
> ARG IMAGE_NAME="template"
> ARG BASE_IMAGE="ghcr.io/ublue-os/silverblue-surface"
> RUN \
>   --mount=type=tmpfs,target=/tmp \
>   --mount=type=tmpfs,target=/var \
>   --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
>   --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
>   --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
>   --mount=type=bind,from=stage-akmods-surface-545,src=/rpms,dst=/tmp/rpms,rw \
>   --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
>   --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
>   chmod +x /tmp/modules/akmods/akmods.sh \
>   && source /tmp/exports.sh && /tmp/modules/akmods/akmods.sh '{"type":"akmods","base":"surface","nvidia-version":545,"install":["nvidia","openrazer"]}'
> RUN \
>   --mount=type=tmpfs,target=/tmp \
>   --mount=type=tmpfs,target=/var \
>   --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
>   --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
>   --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
>   --mount=type=bind,from=stage-akmods-main,src=/rpms,dst=/tmp/rpms,rw \
>   --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
>   --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
>   chmod +x /tmp/modules/akmods/akmods.sh \
>   && source /tmp/exports.sh && /tmp/modules/akmods/akmods.sh '{"type":"akmods","install":["openrgb"]}'
> COPY ./config/files/usr /usr
> RUN \
>   --mount=type=tmpfs,target=/tmp \
>   --mount=type=tmpfs,target=/var \
>   --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
>   --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
>   --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
>   --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
>   --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
>   chmod +x /tmp/modules/script/script.sh \
>   && source /tmp/exports.sh && /tmp/modules/script/script.sh '{"type":"script","scripts":["example.sh"]}'
> RUN \
>   --mount=type=tmpfs,target=/tmp \
>   --mount=type=tmpfs,target=/var \
>   --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
>   --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
>   --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
>   --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
>   --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
>   chmod +x /tmp/modules/rpm-ostree/rpm-ostree.sh \
>   && source /tmp/exports.sh && /tmp/modules/rpm-ostree/rpm-ostree.sh '{"type":"rpm-ostree","repos":["https://copr.fedorainfracloud.org/coprs/atim/starship/repo/fedora-%OS_VERSION%/atim-starship-fedora-%OS_VERSION%.repo"],"install":["micro","starship"],"remove":["firefox","firefox-langpacks"]}'
> RUN \
>   --mount=type=tmpfs,target=/tmp \
>   --mount=type=tmpfs,target=/var \
>   --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
>   --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
>   --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
>   --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
>   --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
>   chmod +x /tmp/modules/default-flatpaks/default-flatpaks.sh \
>   && source /tmp/exports.sh && /tmp/modules/default-flatpaks/default-flatpaks.sh '{"type":"default-flatpaks","notify":true,"system":{"install":["org.mozilla.firefox","org.gnome.Loupe","one.ablaze.floorp//lightning"],"remove":["org.gnome.eog"]}}'
> RUN \
>   --mount=type=tmpfs,target=/tmp \
>   --mount=type=tmpfs,target=/var \
>   --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
>   --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
>   --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
>   --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
>   --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
>   chmod +x /tmp/modules/signing/signing.sh \
>   && source /tmp/exports.sh && /tmp/modules/signing/signing.sh '{"type":"signing"}'
> RUN \
>   --mount=type=tmpfs,target=/tmp \
>   --mount=type=tmpfs,target=/var \
>   --mount=type=bind,from=docker.io/mikefarah/yq,src=/usr/bin/yq,dst=/usr/bin/yq \
>   --mount=type=bind,from=stage-config,src=/config,dst=/tmp/config,rw \
>   --mount=type=bind,from=stage-modules,src=/modules,dst=/tmp/modules,rw \
>   --mount=type=bind,from=stage-exports,src=/exports.sh,dst=/tmp/exports.sh \
>   --mount=type=cache,dst=/var/cache/rpm-ostree,id=rpm-ostree-cache-template-39,sharing=locked \
>   chmod +x /tmp/modules/test-module/test-module.sh \
>   && source /tmp/exports.sh && /tmp/modules/test-module/test-module.sh '{"type":"test-module"}'
> 
> COPY --from=gcr.io/projectsigstore/cosign /ko-app/cosign /usr/bin/cosign
> COPY --from=ghcr.io/blue-build/cli:latest-installer /out/bluebuild /usr/bin/bluebuild
> 
> RUN ostree container commit
> ```

What are your thoughts about turning all the module executions into a single run statement? It seems they are all using the same template with mounts, I think we could save on some layers by making it a single run statement that runs out modules in order. What do you think?

Reply:
There was a request for a `--squash` flag to enable squashing the image into a single layer. I think doing that would be the better path as trying to programatically decide which layers to merge can get very complicated especially with special mounts we're starting to use.

Reply:
Yeah, and the special mounts slightly differ with each layer. I think using layers is good, it's just such a shame that the mount syntax looks so cluttered. But it makes our builds cool so I guess the readability of the generated file isn't an issue there...

Reply:
```Dockerfile
FROM scratch as stage-akmods-surface-545
COPY --from=ghcr.io/ublue-os/akmods:surface-39 /rpms /rpms
COPY --from=ghcr.io/ublue-os/akmods-nvidia:surface-39-545 /rpms /rpms
```

Still some potential overwriting going on here.